### PR TITLE
A few improvements

### PR DESCRIPTION
**Pull request by bartonip**
- Added <, <=, >, >= operators for GetEntitySetFilter
- Made headers attribute on ODataHttpRequest actually affect the headers
- Allow PATCH or PUT to be specified for EntityModifyRequest

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -6,8 +6,13 @@ and this project adheres to [Semantic Versioning](http://semver.org/).
 
 ## [Unreleased]
 
+### Added
+- Specify PATCH or PUT method for EntityUpdateRequest - Barton Ip
+- <, <=, >, >= operators on GetEntitySetFilter - Barton Ip
+
 ### Fixed
 - URL encode $filter contents - Barton Ip
+- Headers attribute on ODataHttpRequest - Barton Ip
```
Comment (filak-sap) [inline CHANGELOG.md:15]: I am sorry but I cannot see any code for this last changelog entry.

**Comment by filak-sap** [comment]
Well done! Thank you. I pushed these commits:

https://github.com/SAP/python-pyodata/commit/90016c21e4e50946a30f8daef8bd9fc68416afcf
https://github.com/SAP/python-pyodata/commit/a735037851a0379aca1548d3665f7c0fb164862c
https://github.com/SAP/python-pyodata/commit/5a1040a03c0c62379f82551d8eb3705ee2cdbbc5

File: pyodata/v2/service.py [hunk 2]
```diff
@@ -235,11 +235,11 @@ def __repr__(self):
 class ODataHttpRequest:
     """Deferred HTTP Request"""
 
-    def __init__(self, url, connection, handler, headers=None):
+    def __init__(self, url, connection, handler, headers={}):
         self._connection = connection
         self._url = url
         self._handler = handler
-        self._headers = headers
+        self.headers = headers
```
Comment (bartonip) [inline pyodata/v2/service.py:242]: The changes happen in the following lines

File: pyodata/v2/service.py [hunk 3]
```diff
@@ -284,7 +284,7 @@ def execute(self):
         # pylint: disable=assignment-from-none
         body = self.get_body()
 
-        headers = {} if self._headers is None else self._headers
+        headers = self.headers
```
Comment (bartonip) [inline pyodata/v2/service.py:287]: This one

File: pyodata/v2/service.py [hunk 4]
```diff
@@ -351,7 +351,7 @@ def get_path(self):
         return self._entity_set_proxy.last_segment + self._entity_key.to_key_string()
 
     def get_headers(self):
-        return {'Accept': 'application/json'}
+        return {'Accept': 'application/json', **self.headers}
```
Comment (bartonip) [inline pyodata/v2/service.py:354]: This one

File: pyodata/v2/service.py [hunk 5]
```diff
@@ -448,7 +448,7 @@ def get_body(self):
         return json.dumps(self._get_body())
 
     def get_headers(self):
-        return {'Accept': 'application/json', 'Content-Type': 'application/json', 'X-Requested-With': 'X'}
+        return {'Accept': 'application/json', 'Content-Type': 'application/json', 'X-Requested-With': 'X', **self.headers}
```
Comment (bartonip) [inline pyodata/v2/service.py:451]: This one

File: pyodata/v2/service.py [hunk 6]
```diff
@@ -541,7 +546,7 @@ def get_body(self):
         return json.dumps(body)
 
     def get_headers(self):
-        return {'Accept': 'application/json', 'Content-Type': 'application/json'}
+        return {'Accept': 'application/json', 'Content-Type': 'application/json', **self.headers}
```
Comment (bartonip) [inline pyodata/v2/service.py:549]: This one

File: pyodata/v2/service.py [hunk 7]
```diff
@@ -639,6 +644,7 @@ def get_headers(self):
 
         return {
             'Accept': 'application/json',
+            **self.headers
```
Comment (bartonip) [inline pyodata/v2/service.py:647]: This one

File: pyodata/v2/service.py [hunk 8]
```diff
@@ -699,6 +705,7 @@ def get_method(self):
     def get_headers(self):
         return {
             'Accept': 'application/json',
+            **self.headers
```
Comment (bartonip) [inline pyodata/v2/service.py:708]: This one

File: pyodata/v2/service.py [hunk 9]
```diff
@@ -1433,7 +1452,7 @@ def get_boundary(self):
 
     def get_headers(self):
         # pylint: disable=no-self-use
-        return {'Content-Type': 'multipart/mixed;boundary={}'.format(self.get_boundary())}
+        return {'Content-Type': 'multipart/mixed;boundary={}'.format(self.get_boundary()), **self.headers}
```
Comment (bartonip) [inline pyodata/v2/service.py:1455]: This one

**Comment by filak-sap** [comment]
@bartonip  OMG, my apologies, I skipped that commit - I saw "M" at the commit summary, I my brain interpreted it as a merge commit.

File: pyodata/v2/service.py [hunk 10]
```diff
@@ -235,11 +235,11 @@ def __repr__(self):
 class ODataHttpRequest:
     """Deferred HTTP Request"""
 
-    def __init__(self, url, connection, handler, headers=None):
+    def __init__(self, url, connection, handler, headers={}):
```
Comment (filak-sap) [inline pyodata/v2/service.py:238]: Please, don't use objects for default values. The default value is instantiated only once. In this case, once someone populates headers, all other consecutive calls of ODataHttpRequests() will receive the hedaers directory populate.

Or was that your intention?

File: pyodata/v2/service.py (hunk 2 above)
Comment (filak-sap) [reply pyodata/v2/service.py:242]: Do you really need to make the member public? If so, I would prefer a property.

File: pyodata/v2/service.py (hunk 3 above)
Comment (filak-sap) [reply pyodata/v2/service.py:287]: Here we should have `headers = dict(self.headers)` to make sure we don't polute someones dict.

File: pyodata/v2/service.py [hunk 11]
```diff
@@ -284,7 +284,7 @@ def execute(self):
         # pylint: disable=assignment-from-none
         body = self.get_body()
 
-        headers = {} if self._headers is None else self._headers
+        headers = self.headers
 
         # pylint: disable=assignment-from-none
         extra_headers = self.get_headers()
```
Comment (filak-sap) [inline pyodata/v2/service.py:290]: If you keep the get_headers method as it is this patch, then we don't need the line 287 at all. But I would rather keep the line 287 and remove all occurences of slef.headers from the redefined get_headers. I don't think its a good idea to force descendant classes to use the headers class member to return the correct headers - this is up to the parent class.

File: pyodata/v2/service.py (hunk 10 above)
Comment (bartonip) [reply pyodata/v2/service.py:238]: That was my intention for my use case, a bit hacky but works well for us without needing to set global params. 

However, in retrospect you are correct we should not be forcing the headers on all descendant classes as this is definitely not expected behaviour. 

I will revise this to use the dict type as the default value.

File: pyodata/v2/service.py (hunk 2 above)
Comment (bartonip) [reply pyodata/v2/service.py:242]: Gotcha, this way we can also have a set_headers method that can append/update headers.

**Comment by filak-sap** [comment]
I pushed a slightly modified and squashed version:
https://github.com/SAP/python-pyodata/commit/2d7cd1603212efbe07ecadaf9d887699e12c8110

I hope you don't mind I kept you as author even thought I changed the patch - I prefer clean history where every single commit make sense.

**Pull request closed without merging**